V2

Anusha20 opened the pull request:
Modified the crawler to added support for distributing tasks across processes and using distributed cache

Anusha20 committed:
Added a basic crawler and maven build file

Anusha20 committed:
Added some comments . Removed override . Moved the test classes to the proper directory structures for mvn

Anusha20 committed:
Read me with details to run the code

Anusha20 committed:
Modified to use external cache and distribute task outside the process

mmcmahon-jana commented on src/main/java/interview/jana/crawler/FindEmailIds.java:
```diff
@@ -0,0 +1,84 @@
+	public static void main(String args[]) {
+
+		initialize(args[0]);
+		String url = "http://" + homeURL;
```
How could you validate the input here to prevent invalid URLs from being parsed?

Anusha20 replied:
Oops, my bad ! I can use API like URLValidator (apache common validator) to validate the input.

mmcmahon-jana commented on src/main/java/interview/jana/crawler/FindEmailIds.java:
```diff
@@ -0,0 +1,84 @@
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		processor.submitForProcessing(url);
```
If the catch block was executed would you still want to make this call?

Anusha20 replied:
good catch . Will move the call inside the block.

mmcmahon-jana commented on src/main/java/interview/jana/crawler/FindEmailIds.java:
```diff
@@ -0,0 +1,84 @@
+		initialize(args[0]);
+		String url = "http://" + homeURL;
+		System.out.println("Found Processing:" + url);
+		UseJedisPubsub(url);
```
Is there a way to verify the Jedis connection was successful and alert the user if not?

Anusha20 replied:
I can getJedisConnection() method to return a false  on JedisException. By alert the user do you mean  make a proper server shutdown. or try another Redis connection ?

mmcmahon-jana replied:
I was thinking of something along the lines of just shutting down app in the case where Redis is not available, giving some kind of message to alert the user as to why it's not running.

mmcmahon-jana commented on src/main/java/interview/jana/crawler/PageCrawler.java:
```diff
@@ -0,0 +1,103 @@
+
+	}
+
+	public void crawlinternalPages(Document doc) {
```
What changes could you make to this method to allow for limiting the depth of the crawling?  For instance, on a site like `web.mit.edu` or `cnn.com`, where the number of pages could be extremely large.

Anusha20 replied:
We can add restriction based on the levels, For example, we can process still level 2 and allow web.mit/engineering/departments and not allow web.mit/engineering/departments/mechanical. For this, we might need to tokenize the URL and find the depth and restrict accordingly.

mmcmahon-jana commented on src/main/java/interview/jana/crawler/PageCrawler.java:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	public Elements getEmailIds(Document doc) {
+		Pattern p = Pattern.compile(emailPattern);
```
Rather than compiling this every time this method is called you could save the Pattern object itself and a static property of the class.

Anusha20 replied:
True , Will make the changes.

Anusha20 committed:
Incorporated changes from review:
- validated URL
- Added filtering based on depth
- implemented shutdown on Jedis connection exception
- Made Pattern static
-
